Added new sound_channels builtin

stefandd opened the pull request:
This adds a new sound command `sound_channels() -> int` by which the number of currently still available channels is obtained. This is useful in situations when more sounds are potentially generated than channels are available, i.e. when for example many enemy ships shoot lasers, potentially filling up the entire mixer channel range. With this command, a Lobster program can check how many channels are still free and decide to not add another sound of a certain type in order to reserve a channel for a more prioritized sound effect (player ship etc.)

aardappel commented:
Doesn't SDL_Mixer handle this automatically already? By default, it has 8 channels, so any sounds attempted to be played while there's already 8 playing will get dropped, which sounds like the right behavior for a simple game.

If anything, we should be calling `Mix_AllocateChannels(16)` after `Mix_OpenAudio` at least, since 8 seems a bit low to me.. that should reduce the chances of sounds dropping too.

If we wanted to make a fancier sound interface, then `play_sound` should take a priority parameter, and then if a sound fails to play because channels are full, it finds the lowest pri sound, kills it if its lower than the current sound, and tries again. That way a "game over" sound would not be dropped because you have 8 bullet sounds playing. In fact I'm surprised SDL_Mixer doesn't already have this functionality.

stefandd commented:
Extended pre-allocated channels to 16; added prioritization of sounds as an argument to the play_xxx functions, and added commands to interrogate the status of a channel (playing, paused, available) etc. The prioritization scheme is that 3 priority levels are supported: low/normal (0), elevated (1), and high (2). Adding a sound will check for a free channel and if none is available will delete the oldest sound effect from the lowest possible priority (including up to the same priority as the caller)

aardappel commented on dev/src/sdlaudiosfxr.cpp:
```diff
@@ -422,14 +424,30 @@ void SDLSoundClose() {
+            }
+        }
+    }
+    if (ch > -1) { // ch should never be -1 at this point anyhow
```
`ch >= 0`

ch might be -1 here if you tried to play something with normal prio, and all channels are taken up by higher prio groups.

aardappel commented on dev/src/audio.cpp:
```diff
@@ -22,31 +22,54 @@ using namespace lobster;
         return Value(ch);
     });
 
+nfr("sound_channels", "", "", "I",
```
I don't think we want this and the following function.. I'd rather sound management be full automatic once you choose good priorities

stefandd replied:
I do agree re: `sound_channels`. Regarding the `sound_status` function, isn't that useful to be able to get the channel (is it paused/playing etc.)?

aardappel replied:
I'm not sure when that would be useful, but left it in for now

aardappel commented on dev/src/sdlaudiosfxr.cpp:
```diff
@@ -422,14 +424,30 @@ void SDLSoundClose() {
+    if (prio < 0) prio = 0;
+    else if (prio > 2) prio = 2;
+    int ch = Mix_GroupAvailable(-1); // is there any free channel?
+    if (ch == -1) { // no free channel?
```
`ch < 0`

aardappel commented on dev/src/sdlaudiosfxr.cpp:
```diff
@@ -422,14 +424,30 @@ void SDLSoundClose() {
+    }
+    if (ch > -1) { // ch should never be -1 at this point anyhow
+        Mix_PlayChannel(ch, snd->chunk.get(), loops);
+        Mix_GroupChannel(ch, prio); // add the channel to the group with tag prio
```
Here we assign a channel to a group.. how does it ever get de-assigned? in particular, any sounds in the highest prio group that finishes playing should immediately be available to the any other prio group.. how does that currently work?

If this gets complicated, I think managing prios without groups would be better

aardappel commented:
Ok, looks good now, thanks!
